receiver/k8sclusterreceiver/e2e_test.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer/consumertest"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	"go.opentelemetry.io/collector/receiver/receivertest"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/k8stest"
+)
+
+const testKubeConfig = "/tmp/kube-config-otelcol-e2e-testing"
+
+const shortenNamesRegex = "([\\D]+)-?.*"
+
+var writeExpected = false
+
+// TestE2E tests the k8s attributes processor with a real k8s cluster.
+// The test requires a prebuilt otelcontribcol image uploaded to a kind k8s cluster defined in
+// `/tmp/kube-config-otelcol-e2e-testing`. Run the following command prior to running the test locally:
+//
+//	kind create cluster --kubeconfig=/tmp/kube-config-otelcol-e2e-testing
+//	make docker-otelcontribcol
+//	KUBECONFIG=/tmp/kube-config-otelcol-e2e-testing kind load docker-image otelcontribcol:latest
+func TestE2E(t *testing.T) {
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", testKubeConfig)
+	require.NoError(t, err)
+	dynamicClient, err := dynamic.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+
+	testID := uuid.NewString()[:8]
+	collectorObjs := k8stest.CreateCollectorObjects(t, dynamicClient, testID)
+
+	defer func() {
+		for _, obj := range append(collectorObjs) {
+			require.NoErrorf(t, k8stest.DeleteObject(dynamicClient, obj), "failed to delete object %s", obj.GetName())
+		}
+	}()
+
+	metricsConsumer := new(consumertest.MetricsSink)
+	wantEntries := 10 // Minimal number of metrics to wait for.
+	waitForData(t, wantEntries, metricsConsumer)
+
+	var expected pmetric.Metrics
+	expectedFile := filepath.Join("testdata", "e2e", "expected.yaml")
+	l := len(metricsConsumer.AllMetrics())
+	if !writeExpected {
+		expected, err = golden.ReadMetrics(expectedFile)
+		require.NoError(t, err)
+	} else {
+		require.GreaterOrEqual(t, l, 1)
+		err := golden.WriteMetrics(t, expectedFile, metricsConsumer.AllMetrics()[l-1])
+		require.NoError(t, err)
+		expected = metricsConsumer.AllMetrics()[l-1]
+	}
+	r, err := regexp.Compile(shortenNamesRegex)
+	require.NoError(t, err)
+	require.NoError(t, pmetrictest.CompareMetrics(expected, metricsConsumer.AllMetrics()[l-1],
+		pmetrictest.IgnoreTimestamp(),
+		pmetrictest.IgnoreStartTimestamp(),
+		pmetrictest.IgnoreMetricValues("k8s.deployment.desired", "k8s.deployment.available", "k8s.container.restarts", "k8s.container.cpu_request", "k8s.container.memory_request", "k8s.container.memory_limit"),
+		pmetrictest.ChangeResourceAttributeValue("k8s.deployment.name", func(value string) string {
+			matches := r.FindStringSubmatch(value)
+			return matches[1]
+		}),
+		pmetrictest.ChangeResourceAttributeValue("k8s.pod.name", func(value string) string {
+			matches := r.FindStringSubmatch(value)
```

Comment:
I don't think this regex will work every time. Pod name can be something like `otelcol-affbacbc-5459b589fd`. In that case, it'll give us a wrong substring. I think it'll be easier just to cut the suffix because the length of it is guaranteed. So for `k8s.pod.name`, it'll be just `return value[:len(value)-20]`. For `k8s.replicaset.name`, it'll be `return value[:len(value)-9].